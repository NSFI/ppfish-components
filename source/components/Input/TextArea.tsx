import * as React from 'react';
import omit from 'omit.js';
import classNames from 'classnames';
import calculateNodeHeight from './calculateNodeHeight';

function onNextFrame(cb: () => void) {
  if (window.requestAnimationFrame) {
    return window.requestAnimationFrame(cb);
  }
  return window.setTimeout(cb, 1);
}

function clearNextFrameAction(nextFrameId: number) {
  if (window.cancelAnimationFrame) {
    window.cancelAnimationFrame(nextFrameId);
  } else {
    window.clearTimeout(nextFrameId);
  }
}

export interface AutoSizeType {
  minRows?: number;
  maxRows?: number;
}

export type HTMLTextareaProps = React.TextareaHTMLAttributes<HTMLTextAreaElement>;

export interface TextAreaProps extends HTMLTextareaProps {
  prefixCls?: string;
  autosize?: boolean | AutoSizeType;
  onPressEnter?: React.KeyboardEventHandler<HTMLTextAreaElement>;
}

export interface TextAreaRef {
  textAreaRef: HTMLTextAreaElement;
  focus: () => void;
  blur: () => void;
  resizeTextarea: () => void;
}

const InternalTextArea: React.ForwardRefRenderFunction<TextAreaRef, TextAreaProps> = (
  props,
  ref,
) => {
  const [textareaStyles, setTextareaStyles] = React.useState<React.CSSProperties>({});
  const nextFrameActionIdRef = React.useRef<number>(null);
  const textAreaRef = React.useRef<HTMLTextAreaElement>(null);

  const resizeTextarea = () => {
    const { autosize } = props;
    const textAreaNode = textAreaRef.current;
    if (!autosize || !textAreaNode) {
      return;
    }
    const minRows = autosize ? (autosize as AutoSizeType).minRows : null;
    const maxRows = autosize ? (autosize as AutoSizeType).maxRows : null;
    const textareaStyles = calculateNodeHeight(textAreaNode, false, minRows, maxRows);
    setTextareaStyles(textareaStyles);
  };

  const getTextAreaClassName = () => {
    const { prefixCls, className, disabled } = props;
    return classNames(prefixCls, className, {
      [`${prefixCls}-disabled`]: disabled,
    });
  };

  const handleTextareaChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    if (!('value' in props)) {
      resizeTextarea();
    }
    const { onChange } = props;
    if (onChange) {
      onChange(e);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    const { onPressEnter, onKeyDown } = props;
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  };

  React.useImperativeHandle(ref, () => ({
    textAreaRef: textAreaRef.current,
    focus: () => {
      textAreaRef.current.focus();
    },
    blur: () => {
      textAreaRef.current.blur();
    },
    resizeTextarea,
  }));

  React.useEffect(() => {
    const nextFrameActionId = nextFrameActionIdRef.current;
    if (nextFrameActionId) {
      clearNextFrameAction(nextFrameActionId);
    }
    nextFrameActionIdRef.current = onNextFrame(resizeTextarea);
  }, [props.value]);

  const otherProps = omit(props, ['prefixCls', 'onPressEnter', 'autosize']);
  const style = {
    ...props.style,
    ...textareaStyles,
  };
  // Fix https://github.com/ant-design/ant-design/issues/6776
  // Make sure it could be reset when using form.getFieldDecorator
  if ('value' in otherProps) {
    otherProps.value = otherProps.value || '';
  }
  return (
    <textarea
      {...otherProps}
      className={getTextAreaClassName()}
      style={style}
      onKeyDown={handleKeyDown}
      onChange={handleTextareaChange}
      ref={textAreaRef}
    />
  );
};

const TextArea = React.forwardRef(InternalTextArea);

TextArea.displayName = 'TextArea';

TextArea.defaultProps = {
  prefixCls: 'fishd-input',
};

export default TextArea;
