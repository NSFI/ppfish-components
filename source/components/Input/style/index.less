@import "../../../assets/css/index";
@import "./mixin";

// Input styles
.@{fishd-prefix}-input {
  .reset-component;
  .input;
}

//== Style for input-group: input with label, with button or dropdown...
.@{fishd-prefix}-input-group {
  .reset-component;
  .input-group(~"@{fishd-prefix}-input");
  &-wrapper {
    display: inline-block;
    vertical-align: top; // https://github.com/ant-design/ant-design/issues/6403
    width: 100%;
  }
}

// Input with affix: prefix or suffix
.@{fishd-prefix}-input-affix-wrapper {
  .reset-component;
  .input-affix-wrapper(~"@{fishd-prefix}-input");

  // https://github.com/ant-design/ant-design/issues/6144
  .@{fishd-prefix}-input {
    min-height: 100%;  // use min-height, assume that no smaller height to override
  }
}

@import "./search-input";
