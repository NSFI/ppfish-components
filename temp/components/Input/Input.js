var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import omit from 'omit.js';
function fixControlledValue(value) {
    if (typeof value === 'undefined' || value === null) {
        return '';
    }
    return value;
}
var Input = /** @class */ (function (_super) {
    __extends(Input, _super);
    function Input() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleKeyDown = function (e) {
            var _a = _this.props, onPressEnter = _a.onPressEnter, onKeyDown = _a.onKeyDown;
            if (e.keyCode === 13 && onPressEnter) {
                onPressEnter(e);
            }
            if (onKeyDown) {
                onKeyDown(e);
            }
        };
        _this.saveInput = function (node) {
            _this.input = node;
        };
        return _this;
    }
    Input.prototype.focus = function () {
        this.input.focus();
    };
    Input.prototype.blur = function () {
        this.input.blur();
    };
    Input.prototype.getInputClassName = function () {
        var _a;
        var _b = this.props, prefixCls = _b.prefixCls, size = _b.size, disabled = _b.disabled;
        return classNames(prefixCls, (_a = {},
            _a[prefixCls + "-sm"] = size === 'small',
            _a[prefixCls + "-lg"] = size === 'large',
            _a[prefixCls + "-disabled"] = disabled,
            _a));
    };
    Input.prototype.renderLabeledInput = function (children) {
        var _a, _b;
        var props = this.props;
        // Not wrap when there is not addons
        if (!props.addonBefore && !props.addonAfter) {
            return children;
        }
        var wrapperClassName = props.prefixCls + "-group";
        var addonClassName = wrapperClassName + "-addon";
        var addonBefore = props.addonBefore ? (React.createElement("span", { className: addonClassName }, props.addonBefore)) : null;
        var addonAfter = props.addonAfter ? (React.createElement("span", { className: addonClassName }, props.addonAfter)) : null;
        var className = classNames(props.prefixCls + "-wrapper", (_a = {},
            _a[wrapperClassName] = addonBefore || addonAfter,
            _a));
        var groupClassName = classNames(props.prefixCls + "-group-wrapper", (_b = {},
            _b[props.prefixCls + "-group-wrapper-sm"] = props.size === 'small',
            _b[props.prefixCls + "-group-wrapper-lg"] = props.size === 'large',
            _b));
        // Need another wrapper for changing display:table to display:inline-block
        // and put style prop in wrapper
        if (addonBefore || addonAfter) {
            return (React.createElement("span", { className: groupClassName, style: props.style },
                React.createElement("span", { className: className },
                    addonBefore,
                    React.cloneElement(children, { style: null }),
                    addonAfter)));
        }
        return (React.createElement("span", { className: className },
            addonBefore,
            children,
            addonAfter));
    };
    Input.prototype.renderLabeledIcon = function (children) {
        var _a;
        var props = this.props;
        if (!('prefix' in props || 'suffix' in props)) {
            return children;
        }
        var prefix = props.prefix ? (React.createElement("span", { className: props.prefixCls + "-prefix" }, props.prefix)) : null;
        var suffix = props.suffix ? (React.createElement("span", { className: props.prefixCls + "-suffix" }, props.suffix)) : null;
        var affixWrapperCls = classNames(props.className, props.prefixCls + "-affix-wrapper", (_a = {},
            _a[props.prefixCls + "-affix-wrapper-sm"] = props.size === 'small',
            _a[props.prefixCls + "-affix-wrapper-lg"] = props.size === 'large',
            _a));
        return (React.createElement("span", { className: affixWrapperCls, style: props.style },
            prefix,
            React.cloneElement(children, {
                style: null,
                className: this.getInputClassName()
            }),
            suffix));
    };
    Input.prototype.renderInput = function () {
        var _a = this.props, value = _a.value, className = _a.className;
        // Fix https://fb.me/react-unknown-prop
        var otherProps = omit(this.props, [
            'prefixCls',
            'onPressEnter',
            'addonBefore',
            'addonAfter',
            'prefix',
            'suffix'
        ]);
        if ('value' in this.props) {
            otherProps.value = fixControlledValue(value);
            // Input elements must be either controlled or uncontrolled,
            // specify either the value prop, or the defaultValue prop, but not both.
            delete otherProps.defaultValue;
        }
        return this.renderLabeledIcon(React.createElement("input", __assign({}, otherProps, { className: classNames(this.getInputClassName(), className), onKeyDown: this.handleKeyDown, ref: this.saveInput })));
    };
    Input.prototype.render = function () {
        return this.renderLabeledInput(this.renderInput());
    };
    Input.defaultProps = {
        prefixCls: 'fishd-input',
        type: 'text',
        disabled: false
    };
    Input.propTypes = {
        type: PropTypes.string,
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        size: PropTypes.oneOf(['small', 'default', 'large']),
        maxLength: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        disabled: PropTypes.bool,
        value: PropTypes.any,
        defaultValue: PropTypes.any,
        className: PropTypes.string,
        addonBefore: PropTypes.node,
        addonAfter: PropTypes.node,
        prefixCls: PropTypes.string,
        autosize: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
        onPressEnter: PropTypes.func,
        onKeyDown: PropTypes.func,
        onKeyUp: PropTypes.func,
        onFocus: PropTypes.func,
        onBlur: PropTypes.func,
        prefix: PropTypes.node,
        suffix: PropTypes.node
    };
    return Input;
}(React.Component));
export default Input;
