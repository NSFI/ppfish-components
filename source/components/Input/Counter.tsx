import * as React from 'react';
import omit from 'omit.js';
import classNames from 'classnames';
import TextArea, { TextAreaProps, TextAreaRef } from './TextArea';
import useControlledState from '../../hooks/useControlledState';

function countValue(value: string) {
  return value.length;
}

export interface CounterProps extends TextAreaProps {
  inputPrefixCls?: string;
  prefixCls?: string;
  limit: number;
  count?: (value: string) => number;
  value?: any;
  defaultValue?: any;
}

export interface CounterRef {
  focus: () => void;
  blur: () => void;
  textarea: TextAreaRef;
}

const InternalCounter: React.ForwardRefRenderFunction<CounterRef, CounterProps> = (props, ref) => {
  const [value, setValue] = useControlledState('', {
    value: props.value,
    defaultValue: props.defaultValue,
  });
  const textareaRef = React.useRef<TextAreaRef>(null);

  const focus = () => {
    textareaRef.current.focus();
  };

  const handleClick = () => {
    focus();
  };

  const handleTextareaChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    const { onChange } = props;
    const textarea = textareaRef.current;
    const textareaValue = textarea && textarea.textAreaRef.value;
    setValue(textareaValue);
    if (onChange) {
      onChange(e);
    }
  };

  const getCount = () => {
    const { count } = props;
    if (!value) {
      return 0;
    }
    // 自定义计数方法
    if (count) {
      return count(String(value));
    }
    return countValue(String(value));
  };

  React.useImperativeHandle(ref, () => ({
    focus,
    blur: () => {
      textareaRef.current.blur();
    },
    textarea: textareaRef.current,
  }));

  const { inputPrefixCls, className, prefixCls, disabled, limit } = props;
  const inputClassName = classNames(className, {
    [`${prefixCls}`]: true,
    [`${inputPrefixCls}-disabled`]: disabled,
  });
  const textareaClassName = classNames(inputPrefixCls, className);
  const otherProps = omit(props, [
    'inputPrefixCls',
    'prefixCls',
    'limit',
    'count',
    'value',
    'onChange',
  ]);
  const total = getCount();
  return (
    <span className={inputClassName} onClick={handleClick}>
      <TextArea
        {...otherProps}
        className={textareaClassName}
        maxLength={limit}
        onChange={handleTextareaChange}
        value={value}
        ref={textareaRef}
      />
      <span className={`${prefixCls}-footer`}>
        <span className={`${prefixCls}-indicator`}>
          {total}/{limit}
        </span>
      </span>
    </span>
  );
};

const Counter = React.forwardRef(InternalCounter);

Counter.displayName = 'Counter';

Counter.defaultProps = {
  inputPrefixCls: 'fishd-input',
  prefixCls: 'fishd-input-counter',
};

export default Counter;
