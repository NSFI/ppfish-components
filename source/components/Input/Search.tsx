import * as React from 'react';
import * as classNames from 'classnames';
import Input, { InputProps } from './Input';
import Icon from '../Icon';
import Button from '../Button';

export interface SearchProps extends InputProps {
  inputPrefixCls?: string;
  onSearch?: (value: string, event?: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLInputElement>) => any;
  enterButton?: boolean | React.ReactNode;
}

export default class Search extends React.Component<SearchProps, any> {
  static defaultProps = {
    inputPrefixCls: 'fishd-input',
    prefixCls: 'fishd-input-search',
    enterButton: false,
  };

  private input: Input;

  onSearch = (e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLInputElement>) => {
    const { onSearch } = this.props;
    if (onSearch) {
      onSearch(this.input.input.value, e);
    }
    this.input.focus();
  }

  focus() {
    this.input.focus();
  }

  blur() {
    this.input.blur();
  }

  saveInput = (node: Input) => {
    this.input = node;
  }

  getButtonOrIcon() {
    const { enterButton, prefixCls, size, disabled } = this.props;
    const enterButtonAsElement = enterButton as React.ReactElement<any>;
    let node;
    if (!enterButton) {
      node = <Icon className={`${prefixCls}-icon`} type="search" key="searchIcon" />;
    } else if (enterButtonAsElement.type === Button || enterButtonAsElement.type === 'button') {
      node = React.cloneElement(enterButtonAsElement, enterButtonAsElement.type === Button ? {
        className: `${prefixCls}-button`,
        size,
      } : {});
    } else {
      node = (
        <Button
          className={`${prefixCls}-button`}
          type="primary"
          size={size}
          disabled={disabled}
          key="enterButton"
        >
          {enterButton === true ? <Icon type="search" /> : enterButton}
        </Button>
      );
    }
    return React.cloneElement(node, {
      onClick: this.onSearch,
    });
  }

  render() {
    const { className, prefixCls, inputPrefixCls, size, suffix,  enterButton, ...others } = this.props;
    delete (others as any).onSearch;
    const buttonOrIcon = this.getButtonOrIcon();
    const searchSuffix = suffix ? [suffix, buttonOrIcon] : buttonOrIcon;
    const inputClassName = classNames(prefixCls, className, {
      [`${prefixCls}-enter-button`]: !!enterButton,
      [`${prefixCls}-${size}`]: !!size,
    });
    return (
      <Input
        onPressEnter={this.onSearch}
        {...others}
        size={size}
        className={inputClassName}
        prefixCls={inputPrefixCls}
        suffix={searchSuffix}
        ref={this.saveInput}
      />
    );
  }
}
