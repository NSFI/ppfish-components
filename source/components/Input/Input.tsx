import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import omit from 'omit.js';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import Counter from './Counter';
import { Omit } from '../../utils/type';

function fixControlledValue<T>(value: T) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }
  return value;
}

export interface InputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size' | 'prefix'> {
  prefixCls?: string;
  className?: string;
  size?: 'large' | 'default' | 'small';
  onPressEnter?: React.KeyboardEventHandler<HTMLInputElement>;
  addonBefore?: React.ReactNode;
  addonAfter?: React.ReactNode;
  prefix?: React.ReactNode;
  suffix?: React.ReactNode;
}

export interface InputRef {
  input: HTMLInputElement;
  focus: () => void;
  blur: () => void;
}

const InternalInput: React.ForwardRefRenderFunction<InputRef, InputProps> = (props, ref) => {
  const inputRef = React.useRef(null);
  React.useImperativeHandle(ref, () => ({
    input: inputRef.current,
    focus: () => {
      inputRef.current.focus();
    },
    blur: () => {
      inputRef.current.blur();
    },
  }));

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { onPressEnter, onKeyDown } = props;
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  };

  const getInputClassName = () => {
    const { prefixCls, size, disabled } = props;
    return classNames(prefixCls, {
      [`${prefixCls}-sm`]: size === 'small',
      [`${prefixCls}-lg`]: size === 'large',
      [`${prefixCls}-disabled`]: disabled,
    });
  };

  const renderLabeledInput = (children: React.ReactElement<any>) => {
    // Not wrap when there is not addons
    if (!props.addonBefore && !props.addonAfter) {
      return children;
    }

    const wrapperClassName = `${props.prefixCls}-group`;
    const addonClassName = `${wrapperClassName}-addon`;
    const addonBefore = props.addonBefore ? (
      <span className={addonClassName}>{props.addonBefore}</span>
    ) : null;

    const addonAfter = props.addonAfter ? (
      <span className={addonClassName}>{props.addonAfter}</span>
    ) : null;

    const className = classNames(`${props.prefixCls}-wrapper`, {
      [wrapperClassName]: addonBefore || addonAfter,
    });

    const groupClassName = classNames(`${props.prefixCls}-group-wrapper`, {
      [`${props.prefixCls}-group-wrapper-sm`]: props.size === 'small',
      [`${props.prefixCls}-group-wrapper-lg`]: props.size === 'large',
    });

    // Need another wrapper for changing display:table to display:inline-block
    // and put style prop in wrapper
    if (addonBefore || addonAfter) {
      return (
        <span className={groupClassName} style={props.style}>
          <span className={className}>
            {addonBefore}
            {React.cloneElement(children, { style: null })}
            {addonAfter}
          </span>
        </span>
      );
    }
    return (
      <span className={className}>
        {addonBefore}
        {children}
        {addonAfter}
      </span>
    );
  };

  const renderLabeledIcon = (children: React.ReactElement<any>) => {
    if (!('prefix' in props || 'suffix' in props)) {
      return children;
    }

    const prefix = props.prefix ? (
      <span className={`${props.prefixCls}-prefix`}>{props.prefix}</span>
    ) : null;

    const suffix = props.suffix ? (
      <span className={`${props.prefixCls}-suffix`}>{props.suffix}</span>
    ) : null;

    const affixWrapperCls = classNames(props.className, `${props.prefixCls}-affix-wrapper`, {
      [`${props.prefixCls}-affix-wrapper-sm`]: props.size === 'small',
      [`${props.prefixCls}-affix-wrapper-lg`]: props.size === 'large',
    });
    return (
      <span className={affixWrapperCls} style={props.style}>
        {prefix}
        {React.cloneElement(children, {
          style: null,
          className: getInputClassName(),
        })}
        {suffix}
      </span>
    );
  };

  const renderInput = () => {
    const { value, className } = props;
    // Fix https://fb.me/react-unknown-prop
    const otherProps = omit(props, [
      'prefixCls',
      'onPressEnter',
      'addonBefore',
      'addonAfter',
      'prefix',
      'suffix',
    ]);

    if ('value' in props) {
      otherProps.value = fixControlledValue(value);
      // Input elements must be either controlled or uncontrolled,
      // specify either the value prop, or the defaultValue prop, but not both.
      delete otherProps.defaultValue;
    }
    return renderLabeledIcon(
      <input
        {...otherProps}
        className={classNames(getInputClassName(), className)}
        onKeyDown={handleKeyDown}
        ref={inputRef}
      />,
    );
  };

  return renderLabeledInput(renderInput());
};

interface InputInterface
  extends React.ForwardRefExoticComponent<InputProps & React.RefAttributes<InputRef>> {
  Group: typeof Group;
  Search: typeof Search;
  TextArea: typeof TextArea;
  Counter: typeof Counter;
}

const Input = React.forwardRef(InternalInput) as InputInterface;

Input.displayName = 'Input';

Input.defaultProps = {
  prefixCls: 'fishd-input',
  type: 'text',
  disabled: false,
};

Input.propTypes = {
  type: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  size: PropTypes.oneOf(['small', 'default', 'large']),
  maxLength: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  disabled: PropTypes.bool,
  value: PropTypes.any,
  defaultValue: PropTypes.any,
  className: PropTypes.string,
  addonBefore: PropTypes.node,
  addonAfter: PropTypes.node,
  prefixCls: PropTypes.string,
  onPressEnter: PropTypes.func,
  onKeyDown: PropTypes.func,
  onKeyUp: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  prefix: PropTypes.node,
  suffix: PropTypes.node,
};

export default Input;
