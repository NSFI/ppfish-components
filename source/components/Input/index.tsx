import Input from './Input';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import './style/index.less';

export {InputProps} from './Input';
export {GroupProps} from './Group';
export {SearchProps} from './Search';
export {TextAreaProps} from './TextArea';

Input.Group = Group;
Input.Search = Search;
Input.TextArea = TextArea;
export default Input;
