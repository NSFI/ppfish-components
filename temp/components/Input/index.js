import Input from './Input';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import Counter from './Counter';
import './style/index.less';
Input.Group = Group;
Input.Search = Search;
Input.TextArea = TextArea;
Input.Counter = Counter;
export default Input;
