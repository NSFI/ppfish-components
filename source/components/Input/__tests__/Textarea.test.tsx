import React from 'react';
import { mount } from 'enzyme';
import Input from '../index';
import { act } from 'react-dom/test-utils';

const globalTimeout = global.setTimeout;
const sleep = async (timeout = 0) => {
  await act(async () => {
    await new Promise(resolve => globalTimeout(resolve, timeout));
  });
};

const { TextArea, Counter } = Input;

describe('TextArea', () => {
  const originalGetComputedStyle = window.getComputedStyle;
  let scrollHeightCalledCount = 1;
  const mockScrollHeightGet = jest.spyOn(HTMLElement.prototype, 'scrollHeight', 'get');
  beforeAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: node => ({
        getPropertyValue: prop => {
          if (prop === 'box-sizing') {
            return originalGetComputedStyle(node)[prop] || 'border-box';
          }
          return originalGetComputedStyle(node)[prop];
        },
      }),
    });
    mockScrollHeightGet.mockImplementation(() => {
      scrollHeightCalledCount += 1;
      return scrollHeightCalledCount * 20;
    });
  });

  afterAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: originalGetComputedStyle,
    });
    mockScrollHeightGet.mockRestore();
  });

  it('should support focus() blur() resizeTextarea(), and can get textarea element', () => {
    const ref = React.createRef<any>();
    mount(<TextArea ref={ref} />);
    ref.current.focus();
    ref.current.blur();
    ref.current.resizeTextarea();
    expect(ref.current.textAreaRef instanceof HTMLTextAreaElement).toBe(true);
  });

  it('should auto calculate height according to content length', async () => {
    const wrapper = mount(
      <TextArea value="" readOnly autosize={{ minRows: 2, maxRows: 6 }} wrap="off" />,
    );
    wrapper.setProps({ value: '1111\n2222\n3333' });
    await sleep(0);
    const prevHeight = wrapper.getDOMNode<HTMLTextAreaElement>().style.height;

    wrapper.setProps({ value: '1111' });
    await sleep(0);
    const currHeight = wrapper.getDOMNode<HTMLTextAreaElement>().style.height;

    expect(prevHeight).not.toBe(currHeight);
  });

  it('should support onPressEnter and onKeyDown', () => {
    const fakeHandleKeyDown = jest.fn();
    const fakeHandlePressEnter = jest.fn();
    const wrapper = mount(
      <TextArea onKeyDown={fakeHandleKeyDown} onPressEnter={fakeHandlePressEnter} />,
    );
    /** KeyCode 65 is A */
    wrapper.find('textarea').simulate('keydown', { keyCode: 65 });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(1);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(0);

    /** KeyCode 13 is Enter */
    wrapper.find('textarea').simulate('keydown', { keyCode: 13 });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(2);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(1);
  });

  it('should support disabled', () => {
    const wrapper = mount(<TextArea disabled />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  describe('maxLength', () => {
    it('should support maxLength', () => {
      const wrapper = mount(<TextArea maxLength={10} />);
      expect(wrapper.render()).toMatchSnapshot();
    });

    it('maxLength should not block control', () => {
      const wrapper = mount(<TextArea maxLength={1} value="light" />);
      expect(wrapper.find('textarea').props().value).toEqual('light');
    });

    it('should exceed maxLength when use IME', () => {
      const onChange = jest.fn();

      const wrapper = mount(<TextArea maxLength={1} onChange={onChange} />);
      // mock 输入法
      wrapper.find('textarea').simulate('compositionStart');
      wrapper.find('textarea').simulate('change', { target: { value: 'zhu' } });
      wrapper.find('textarea').simulate('compositionEnd', { currentTarget: { value: '竹' } });
      wrapper.find('textarea').simulate('change', { target: { value: '竹' } });

      expect(onChange).toHaveBeenLastCalledWith(
        expect.objectContaining({ target: expect.objectContaining({ value: '竹' }) }),
      );
    });
  });

  it('when prop value not in this.props, resizeTextarea should be called', async () => {
    const wrapper = mount(<TextArea aria-label="textarea" autosize />);

    const prevHeight = wrapper.getDOMNode<HTMLTextAreaElement>().style.height;
    wrapper.find('textarea').simulate('change', {
      target: {
        value: 'test',
      },
    });
    await sleep(0);
    const currHeight = wrapper.getDOMNode<HTMLTextAreaElement>().style.height;
    expect(prevHeight).not.toBe(currHeight);
  });

  it('handleKeyDown', () => {
    const onPressEnter = jest.fn();
    const onKeyDown = jest.fn();
    const wrapper = mount(
      <TextArea onPressEnter={onPressEnter} onKeyDown={onKeyDown} aria-label="textarea" />,
    );
    wrapper.find('textarea').simulate('keydown', { keyCode: 13 });
    expect(onPressEnter).toHaveBeenCalled();
    expect(onKeyDown).toHaveBeenCalled();
  });

  it('should display defaultValue when value is undefined', () => {
    const wrapper = mount(<TextArea defaultValue="Light" />);
    const value = wrapper.getDOMNode<HTMLTextAreaElement>().value;
    expect(value).toBe('Light');
  });

  describe('Counter', () => {
    it('click outside should also get focus', () => {
      const ref = React.createRef<any>();
      const wrapper = mount(<Counter value={111} limit={100} ref={ref} />);
      const onFocus = jest.spyOn(ref.current.textarea, 'focus');
      wrapper.find('.fishd-input-counter').simulate('click');
      expect(onFocus).toHaveBeenCalled();
    });

    it('should support focus() blur() textarea', () => {
      const ref = React.createRef<any>();
      mount(<Counter limit={100} ref={ref} />);
      ref.current.focus();
      ref.current.blur();
      ref.current.textarea.focus();
      ref.current.textarea.blur();
      ref.current.textarea.resizeTextarea();
    });
  });
});
