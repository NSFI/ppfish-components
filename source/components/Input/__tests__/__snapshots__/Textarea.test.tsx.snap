// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea maxLength should support maxLength 1`] = `
<textarea
  class="fishd-input"
  maxlength="10"
/>
`;

exports[`TextArea should support disabled 1`] = `
<textarea
  class="fishd-input fishd-input-disabled"
  disabled=""
/>
`;
